### ISSUE TITLE
EvTable doesn't support Chinese.

### ISSUE THREAD

**Issue by Su9Tail** [evennia/evennia#1522]
When I use `evtable.EvTable("测试"，"测试", table=[[], []])`like this, It didn't display well.

The result is kind like
```
+------+------+
|      |      |
|      |      |
|  测试  |  测试  |
|      |      |
|      |      |
+~~~~~~+~~~~~~+
|      |      |
|      |      |
|  50  |  50  |
|      |      |
|      |      |
+------+------+
```

**Comment by chainsol** [comment]
This one is relatively easy to explain.
```
len("测试") == 6
```
Even though it takes up only two spaces.

**Comment by Griatch** [comment]
Yeah, not sure how to go about this one, honestly.

**Comment by chainsol** [comment]
@Griatch, where do you think the processing for grabbing the width of the various EvCells is? I'm ripping through the code to try to chase that down - it might be the cause.

To correct my earlier example, by the way, if one uses `u'测试'` the length is correct - 2 - but Python does in fact not have a good way to check the visible length of a string.

**Comment by chainsol** [comment]
On doing more research, I'm of the impression that this has to be a won't fix issue. [This link](http://denisbider.blogspot.com/2015/09/when-monospace-fonts-arent-unicode.html) explains even more about the issue - basically fonts don't really display non-Western characters all that well, even in monospace/fixed width fonts. There doesn't seem to be a good solution I've been able to turn up. Evennia is handling this correctly - it's using Unicode throughout, so the reported length of the string is correct - this is merely a display issue.

**Comment by busbyra** [comment]
If you can track down how it is read into EvCells, you can modify EvCells
to be dynamic and adjust the width accordingly, I think.

**Comment by chainsol** [comment]
The problem is that as far as EvCells are concerned, 测 is one visible character. You'd think using a monospace font would correctly deal with non-Western characters, but it _doesn't_.
```
_ _ _
1 2 3
测 测 测
```

**Comment by busbyra** [comment]
No, I get what the issue is. I'm trying to think on how this can be fixed
in a dynamic or pseudo-dynamic way, like detecting character type and
adjusting the width to be correct when stored/printed.



On Dec 7, 2017 12:12 PM, "Nicholas" <notifications@github.com> wrote:

The problem is that as far as EvCells are concerned, 测 is one visible
character. You'd think using a monospace font would correctly deal with
non-Western characters, but it *doesn't*.

_ _ _
1 2 3
测 测 测

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/evennia/evennia/issues/1522#issuecomment-350033664>,
or mute
the thread

**Comment by chainsol** [comment]
Each font would deal with it differently, I don't think we can or should try fixing this. If you can find a monospace font that correctly deals with non-Western characters, that would fix it. The link I gave also tells us that you _could_ fix this dynamically, but only for certain characters - many characters end up being "ambiguous" width. @Griatch any thoughts here?

**Comment by Griatch** [comment]
I think the issue of *visible* font width is not something we can deal with reliably. You would need to use the right font in the web client as @chainsol suggests. Also telnet clients often allows for changing the used font so users could be instructed to use the corrected one. I added a reference to this discussion to the [coding FAQ](https://github.com/evennia/evennia/wiki/Coding-FAQ) for future reference. 
Closing for now. re-open if some new brilliant idea about how to handle this dynamically emerges. :)
